src/test/java/com/cloudbees/jenkins/plugins/advisor/BundleUploadTest.java

```diff
@@ -0,0 +1,151 @@
+  private static StreamHandler customLogHandler;
+
+  @Rule
+  public JenkinsRule j = new JenkinsRule();
+  @Rule
+  public WireMockRule wireMockRule = new WireMockRule(wireMockConfig().port(9999));
+
+
+  @Before
+  public void attachLogCapture() {
+    logCapturingStream = new ByteArrayOutputStream();
+    Handler[] handlers = LOG.getParent().getHandlers();
+    customLogHandler = new StreamHandler(logCapturingStream, handlers[0].getFormatter());
+    LOG.addHandler(customLogHandler);
+    LOG.setLevel(Level.FINEST);
+  }
+
+  private static String getTestCapturedLog() throws IOException {
+    customLogHandler.flush();
+    return logCapturingStream.toString();
+  }
+
+  @WithTimeout(30)
+  @Test
+  public void execute() throws Exception {
+    BundleUpload subject = j.getInstance().getExtensionList(BundleUpload.class).get(0);
+
+    AdvisorGlobalConfiguration config = AdvisorGlobalConfiguration.getInstance();
+    config.setEmail(TEST_EMAIL);
+    config.setPassword(Secret.fromString(TEST_PASSWORD));
+    config.setValid(true);
+
+    Gson gson = new Gson();
+    stubFor(post(urlEqualTo("/login"))
+        .withHeader("Content-Type", WireMock.equalTo("application/json"))
+        .withRequestBody(equalToJson(gson.toJson(new AccountCredentials(TEST_EMAIL, TEST_PASSWORD,null, -1, null, null, null))))
+        .willReturn(aResponse()
+            .withStatus(200)
+            .withHeader("Authorization", "Bearer " + TEST_TOKEN)));
+
+    stubFor(post(urlEqualTo(format("/api/users/%s/upload/%s", TEST_EMAIL, j.getInstance().getLegacyInstanceId())))
+        .withHeader("Authorization", WireMock.equalTo(TEST_TOKEN))
+        .willReturn(aResponse()
+            .withStatus(200)));
+
+    subject.execute(NOOP_LISTENER);
+
+    // hack as wiremock doesn't seem to handle async requests
+    while(wireMockRule.getAllServeEvents().size() < 2) {
+      Thread.sleep(1000L);
+    }
+
+    verify(postRequestedFor(urlEqualTo("/login"))
+        .withHeader("Content-Type", WireMock.equalTo("application/json")));
+
+    verify(postRequestedFor(urlEqualTo(format("/api/users/%s/upload/%s", TEST_EMAIL, j.getInstance().getLegacyInstanceId())))
+        .withHeader("Content-Type", WireMock.containing("multipart/form-data")));
+  }
+
+  @Test
+  public void execute_pluginDisabled() throws Exception {
+    BundleUpload subject = j.getInstance().getExtensionList(BundleUpload.class).get(0);
+    j.getPluginManager().getPlugin(AdvisorGlobalConfiguration.PLUGIN_NAME).disable();
+
+    stubFor(any(anyUrl()));
+
+    subject.execute(NOOP_LISTENER);
+
+    verify(0, anyRequestedFor(anyUrl()));
+  }
+
+  @Test
+  public void execute_isNotValid() throws Exception {
+    BundleUpload subject = j.getInstance().getExtensionList(BundleUpload.class).get(0);
+    AdvisorGlobalConfiguration config = AdvisorGlobalConfiguration.getInstance();
+    config.setValid(false);
+
+    stubFor(any(anyUrl()));
+
+    subject.execute(NOOP_LISTENER);
+
+    verify(0, anyRequestedFor(anyUrl()));
```

Comment:
Seems weird to have a test for this as it's literally just returning a 404 from the server?

Reply:
This is testing that we don't hit the server if "valid" check box is not selected - https://github.com/jenkinsci/cloudbees-jenkins-advisor-plugin/blob/master/src/main/java/com/cloudbees/jenkins/plugins/advisor/BundleUpload.java#L61